scanner/enricher/csaf/fetcher.go

```diff
@@ -0,0 +1,417 @@
+package csaf
+
+import (
+	"archive/tar"
+	"bytes"
+	"context"
+	"encoding/csv"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/klauspost/compress/snappy"
+	"github.com/quay/claircore/libvuln/driver"
+	"github.com/quay/claircore/pkg/tmp"
+	"github.com/quay/zlog"
+	"github.com/stackrox/rox/pkg/utils"
+	"github.com/stackrox/rox/scanner/enricher/csaf/internal/zreader"
+)
+
+var (
+	// compressedFileTimeout matches Claircore's VEX https://github.com/quay/claircore/blob/v1.5.34/rhel/vex/fetcher.go.
+	compressedFileTimeout = 2 * time.Minute
+)
+
+// fingerprint is used to track the state of the changes.csv and deletions.csv endpoints.
+//
+// The spec (https://www.rfc-editor.org/rfc/rfc9110.html#name-etag) mentions
+// that there is no need for the client to be aware of how each entity tag
+// is constructed, however, it mentions that servers should avoid backslashes.
+// Hence, the `\` character is used as a separator when stringifying.
+type fingerprint struct {
+	changesEtag, deletionsEtag string
+	requestTime                time.Time
+	version                    string
+}
+
+// parseFingerprint takes a generic driver.Fingerprint and creates a vex.fingerprint.
+// The string format saved in the DB is returned by the fingerprint.String() method.
+func parseFingerprint(in driver.Fingerprint) (*fingerprint, error) {
+	fp := string(in)
+	if fp == "" {
+		return &fingerprint{}, nil
+	}
+	f := strings.Split(fp, `\`)
+	if len(f) != 4 {
+		return nil, errors.New("could not parse fingerprint")
+	}
+	rt, err := time.Parse(time.RFC3339, f[2])
+	if err != nil {
+		return nil, fmt.Errorf("could not parse fingerprint's requestTime: %w", err)
+	}
+	return &fingerprint{
+		changesEtag:   f[0],
+		deletionsEtag: f[1],
+		requestTime:   rt,
+		version:       f[3],
+	}, nil
+}
+
+// String represents a fingerprint in string format with `\` acting as the delimiter.
+func (fp *fingerprint) String() string {
+	return fp.changesEtag + `\` + fp.deletionsEtag + `\` + fp.requestTime.Format(time.RFC3339) + `\` + fp.version
+}
+
+// FetchEnrichment implements driver.EnrichmentUpdater.
```

Comment:
nit: Add description of what this is doing, inputs, outputs, files, etc.

Reply:
i was hoping to avoid doing this because I mostly copied this over from Claircore, but I'll add some more info